Cleaner logging tests

mxgrey opened the pull request:
This PR takes advantage of `rcutils_logging_set_output_handler` to override the way log outputs are published within the logging test function. We prevent the messages from being printed and instead collect them in way that the test can access them and validate what is being produced.

This has two big advantages:
* We no longer get false "Error" and "Warning" messages printing out to the terminal when running these tests
* We can explicitly validate that the logger is behaving the way we expect

This branch builds directly off of #7, so that should be merged before this is reviewed.

mxgrey committed:
Add a .repos file for jazzy (#425)

Signed-off-by: Michael X. Grey <grey@openrobotics.org>
Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Introduce Logger and LogParams

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Add test for arbitrary string logger name

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
merge with main

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Fix merge

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Add safety comment

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Fix formatting

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Allow &str to implement AsLogParams instead of &&str

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Add safety comment

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Add test for function!() macro

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Replace all uses of LazyLock with OnceLock to be compatible with 1.75

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Fix formatting

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Investigating how to set a custom logging handler

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Comply with Rust naming conventions

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Comply with clippy suggestions

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
merge

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Putting log handling into its own module

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Fix formatting with ToLogParams

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Merge branch 'logging_params' into clean_logging_tests

mxgrey committed:
Able to override the log handling

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Try to help cargo docs resolve macro links

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Merge branch 'logging_params' into clean_logging_tests

mxgrey committed:
Cleaner log testing

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Fork expectations based on distro version

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Account for rolling distro

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Merge remote-tracking branch 'mxgrey/logging_params' into clean_logging_tests

mxgrey committed:
Fix clippy warnings

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey commented:
I just realized a major issue with some of the changes in this PR.

When you specify a custom log output handler function to rcutils, your handler function will be given the format string (`"%s"`) along with the variadic arguments that the user passed in to fill into the format string. 

The variadic arguments are pretty much indecipherable from the Rust side, so in order to make sense of the log messages on the rclrs side, I skipped passing `("%s", msg)` to rcutils and instead simplified it to just passing in `msg` with no variadic arguments. 

This is fine in our tests since we can choose to just look at the message and ignore the variadic arguments. However this will be a problem when we're not testing because it's possible for someone to pass a string with a `%` character into their log message, and when rcl processes that according to its normal formatting rules, it can result in a segmentation fault or other undefined behavior. 

I'll need to tweak the PR to have slightly different behavior when compiled for tests vs compiled for real. A better solution would be to find a way to format a string from Rust given C variadic arguments, but I expect that to be an unreasonably difficult thing to do.

mxgrey committed:
Fix function names generated by logger

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

mxgrey committed:
Fix format string memory vulnerability while still being able to run logging tests

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

mxgrey commented:
The safety issue is fixed by [89809d1](https://github.com/geoff-imdex/ros2_rust/pull/11/commits/89809d15980140c66ff6910db542a615dcb74e9f), so this PR should be good to go.

mxgrey committed:
Fix formatting

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

mxgrey committed:
More robust detection of using custom log handler

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

mxgrey committed:
Remember to reset USING_CUSTOM_HANDLER when the logger gets reset

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

mxgrey committed:
Merge remote-tracking branch 'geoff-imdex/main' into clean_logging_tests

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Merge branch 'clean_logging_tests' of ssh://github.com/mxgrey/ros2_rust into clean_logging_tests

mxgrey committed:
Use cfg instead of suppressing warning

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Fix errors in documentation

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

geoff-imdex commented on rclrs/src/logging/logging_configuration.rs:
```diff
@@ -55,3 +49,201 @@ impl Drop for LoggingLifecycle {
+    /// Reset the logging output handler to the default one
+    pub(crate) fn reset_logging_output_handler() {
+        let _lifecycle = ENTITY_LIFECYCLE_MUTEX.lock().unwrap();
+        unsafe {
```
Sorry, very pedantic, but would it be possible to add the "safety" comment?

mxgrey replied:
That's definitely not pedantic or something to apologize for, thanks for catching it. I wish clippy would flag missing safety comments.. it becomes easy to forget when dealing with language bindings that need dozens of `unsafe` blocks.

Fixed: [8c294d1](https://github.com/geoff-imdex/ros2_rust/pull/11/commits/8c294d1bce7efa35c779232d278c344a023fe52e)

geoff-imdex reviewed (commented):
As always, lots of great stuff (definitely helping my Rust learnings!).  One extremely pedantic comment for adding in the standard `safety` stuff.

mxgrey committed:
Add safety comment

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey committed:
Use latest stable Rust CI + Fix Test Errors (#420)

* Use latest stable Rust CI

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

* Fix quotation in doc

Signed-off-by: Michael X. Grey <grey@openrobotics.org>

* Fix serde feature for vendored messages

Signed-off-by: Michael X. Grey <grey@openrobotics.org>

* Fix formatting of lists in docs

Signed-off-by: Michael X. Grey <grey@openrobotics.org>

* Update minimum supported Rust version based on the currently used language features

Signed-off-by: Michael X. Grey <grey@openrobotics.org>

* Conform to clippy style guide

Signed-off-by: Michael X. Grey <grey@openrobotics.org>

* Add rust toolchain as a matrix dimension

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

* Bump declaration of minimum supported rust version

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

* Limit the scheduled runs to once a week

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

* Define separate stable and minimal workflows because matrix does not work with reusable actions

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>

* Reduce minimum version to 1.75 to target Noble

Signed-off-by: Michael X. Grey <grey@openrobotics.org>

---------

Signed-off-by: Michael X. Grey <mxgrey@intrinsic.ai>
Signed-off-by: Michael X. Grey <grey@openrobotics.org>

nwn committed:
Update vendored interface packages (#423)

* Update rclrs vendor_interfaces.py script

This updates the vendor_interfaces.py script to also vendor in the
action_msgs and unique_identifier_msgs packages. The script is modified
to always use a list of package names rather than hard-coding the
package names everywhere.

* Silence certain clippy lints on generated code

In case a user enforces clippy linting on these generated packages,
silence expected warnings. This is already the case in rclrs, but should
be applied directly to the generated packages for the sake of downstream
users.

The `clippy::derive_partial_eq_without_eq` lint was already being
disabled for the packages vendored by rclrs, but is now moved to the
rosidl_generator_rs template instead. This is necessary since we always
derive the PartialEq trait, but can't necessary derive Eq, and so don't.

The `clippy::upper_case_acronyms` is new and was added to account for
message type names being upper-case acrynyms, like
unique_identifier_msgs::msg::UUID.

* Update vendored message packages

This updates the message packages vendored under the rclrs `vendor`
module. The vendor_interfaces.py script was used for reproducibility.

The existing packages – rcl_interfaces, rosgraph_msgs, and
builtin_interfaces – are only modified in that they now use the
std::ffi::c_void naming for c_void instead of the std::os::raw::c_void
alias and disable certain clippy lints.

The action_msgs and unique_identifier_msgs packages are newly added to
enable adding action support to rclrs. action_msgs is needed for
interaction with action goal info and statuses, and has a dependency on
unique_identifier_msgs.

mxgrey committed:
Change API to use throttle

Signed-off-by: Michael X. Grey <greyxmike@gmail.com>

mxgrey commented:
Based on some discussions in the working group, I've updated `interval` to be called `throttle` instead, aligning the terminology with the other ROS client libraries.

mxgrey committed:
Merge remote-tracking branch 'origin/main' into clean_logging_tests

mxgrey commented:
I've also merged in the latest `main` branch from upstream. I think this PR should be good to go. Hopefully this will be the last one needed for the maintainer review.
